ir/instr.cpp

```diff
@@ -1367,6 +1369,38 @@ unique_ptr<Instr> InsertValue::dup(const string &suffix) const {
   return ret;
 }
 
+DEFINE_AS_RETZERO(FnCall, getMaxAllocSize);
+DEFINE_AS_RETZERO(FnCall, getMaxGEPOffset);
+
+bool FnCall::canFree() const {
+  return !getAttributes().has(FnAttrs::NoFree);
+}
+
+uint64_t FnCall::getMaxAccessSize() const {
+  uint64_t sz = attrs.has(FnAttrs::Dereferenceable) ? attrs.getDerefBytes() : 0;
+  for (auto &arg : args) {
+    if (arg.second.has(ParamAttrs::Dereferenceable))
+      sz = max(sz, arg.second.getDerefBytes());
+  }
+  return sz;
+}
+
+FnCall::ByteAccessInfo FnCall::getByteAccessInfo() const {
+  bool has_deref = getAttributes().has(FnAttrs::Dereferenceable);
+  if (!has_deref) {
+    for (auto &arg : args)
+      if (arg.second.has(ParamAttrs::Dereferenceable)) {
+        has_deref = true;
+        break;
+      }
+  }
+  if (!has_deref)
+    // No dereferenceable attribute
+    return {};
+  // dereferenceable(n) does not guarantee that the pointer is n-byte aligned
+  return ByteAccessInfo::intOnly(1);
```

Comment:
this is a pessimizing things a bit:
 1) it forces int bytes to exist. We should have a "any" return to avoid this. I think there are other fns that could use the same (e.g. memcmp)
 2) it brings all bits_byte down to 8. We can't have all functions with fn calls to have 8-bit bytes. That's a big perf impact. We need a way around this.

Reply:
1. For any byte this should be an easy change.

2. We can check whether all path to dereferenceable requires some alignment of the pointer, but this requires flow-sensitive analysis.
I guess when LLVM infers `dereferenceable` from preceding memory accesses, it will also attach `align` tags. It will naturally solve the problem. When `align` is supported, this should be fixed (which should be after paper submission). What do you think?

Reply:
hopefully it shouldn't need any data-flow analysis, just a bit of pattern matching maybe? It doesn't have to be perfect; just be  good for the 99% cases, and give up elsewhere I guess. But let's think about this after the submission, yes :)